Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py

```diff
@@ -26,82 +63,420 @@ def __init__(self, parent):
+
+
+#
+# VectorToScalarVolumeLogic
+#
+
+class VectorToScalarVolumeLogic(ScriptedLoadableModuleLogic):
+  """
+  Implement the logic to compute the transform from vector to scalar.
+  It is stateless, with the run function getting inputs and setting outputs.
+  """
+
+  LUMINANCE = 'LUMINANCE'
+  AVERAGE = 'AVERAGE'
+  SINGLE_COMPONENT = 'SINGLE_COMPONENT'
+
+  CONVERSION_METHODS = (LUMINANCE, AVERAGE, SINGLE_COMPONENT)
+
+
+  def __init__(self, parent = None):
+    ScriptedLoadableModuleLogic.__init__(self, parent)
+
+  @staticmethod
+  def isValidInputOutputData(inputVolumeNode, outputVolumeNode, conversionMethod, componentToExtract):
+    """
+    Validate parameters using the parameterNode.
+    """
+    #
+    # Checking input/output consistency.
+    #
+    if not inputVolumeNode:
+      msg = 'no input volume node defined'
+      logging.debug("isValidInputOutputData failed: %s" % msg)
+      return False, msg
+    if not outputVolumeNode:
+      msg = 'no output volume node defined'
+      logging.debug("isValidInputOutputData failed: %s" % msg)
+      return False, msg
+    if inputVolumeNode.GetID() == outputVolumeNode.GetID():
+      msg = 'input and output volume is the same. ' \
+            'Create a new volume for output to avoid this error.'
+      logging.debug("isValidInputOutputData failed: %s" % msg)
+      return False, msg
+
+    #
+    # Checking based on method selected
+    #
+    if conversionMethod not in (VectorToScalarVolumeLogic.SINGLE_COMPONENT,
+                                VectorToScalarVolumeLogic.LUMINANCE,
+                                VectorToScalarVolumeLogic.AVERAGE):
+      msg = 'conversionMethod %s unrecognized.' % conversionMethod
+      logging.debug("isValidInputOutputData failed: %s" % msg)
+      return False, msg
+
+    inputImage = inputVolumeNode.GetImageData()
+    nrbOfComponents = inputImage.GetNumberOfScalarComponents()
+
+    # SINGLE_COMPONENT: Check that input has enough components for the given componentToExtract
+    if conversionMethod == VectorToScalarVolumeLogic.SINGLE_COMPONENT:
+      # componentToExtract is an index with valid values in the range: [0, nrbOfComponents-1]
+      if not(0 <= componentToExtract < nrbOfComponents):
+        msg = 'componentToExtract %d is invalid. Image has only %d components.' % (componentToExtract, nrbOfComponents)
+        logging.debug("isValidInputOutputData failed: %s" % msg)
+        return False, msg
+
+    # LUMINANCE: Check that input vector has at least three components.
+    if conversionMethod == VectorToScalarVolumeLogic.LUMINANCE:
+      if nrbOfComponents < 3:
+        msg = 'input has only %d components but requires ' \
+              'at least 3 components for luminance conversion.' % nrbOfComponents
+        logging.debug("isValidInputOutputData failed: %s" % msg)
+        return False, msg
+
+    return True, None
+
+
+  def run(self, inputVolumeNode, outputVolumeNode, conversionMethod, componentToExtract):
```

Comment:
If we use a parameter node then the primary method for running the processing should take a parameter node as input.

In addition to that method, you may add a convenience method, which takes, inputVolume, outputVolume, conversionMethod, etc., creates a parameter node and calls the primary method with that.

Reply:
Done